corehq/ex-submodules/pillowtop/pillow/interface.py

```diff
@@ -100,29 +100,94 @@ def run(self):
         self.process_changes(since=self.get_last_checkpoint_sequence(), forever=True)
 
     def _update_checkpoint(self, change, context):
-        if change:
+        if change and context:
             updated = self.fire_change_processed_event(change, context)
         else:
             updated = self.checkpoint.touch(min_interval=CHECKPOINT_MIN_WAIT)
         if updated:
             self._record_checkpoint_in_datadog()
 
+    @property
+    def _should_process_in_chunks(self):
+        # chunked processing is only supported for pillows with single processor
+        return (len(self.processors) == 1
+            and getattr(self.processors[0], 'processor_chunk_size', 0) > 0)
+
     def process_changes(self, since, forever):
         """
-        Process changes from the changes stream.
+        Handle changes stream from the changes stream and pass it to processors
         """
         context = PillowRuntimeContext(changes_seen=0)
         try:
+            changes_chunk = []
             for change in self.get_change_feed().iter_changes(since=since or None, forever=forever):
                 if change:
                     context.changes_seen += 1
-                    self.process_with_error_handling(change, context)
+                    if self._should_process_in_chunks:
+                        changes_chunk.append(change)
+                        if len(changes_chunk) == self.processors[0].processor_chunk_size:
```

Comment:
I thought that the tmeout handling is taken care of in `self.get_change_feed().iter_changes`, isn't it so?

Reply:
@emord means a timeout for setting the checkpoint. You can see how we currently handle that here: https://github.com/dimagi/commcare-hq/blob/232b6696412e4fbbd9df4b05f55b8738af7efe33/corehq/ex-submodules/pillowtop/checkpoints/manager.py#L133

Reply:
The manager example deals with current iteration and previous iteration (so snapshots are available to compare). Here we need to deal with current iteration and an expected next iteration that might be delayed, so I don't see an easy way without using some sort of parallel polling.

The only graceful way of doing this might be to have  `self.get_change_feed().iter_changes` re-raise `ConsumerTimeout` error and processing remaining chunk and restart `process_changes` (similar to how `PillowCheckpointReset` is handled currently. 

Is there a better way?

Reply:
You'll have keep track of the last time the processing was called:
```
if len(changes_chunk) >= self.processors[0].processor_chunk_size or datetime.utcnow() - last_process_time > timedelta(minutes=X):
     last_process_time = datetime.utcnow()
     ....
```

Reply:
This works only if the next change comes after a certain `MAX_WAIT_TIME`, since this condition  gets evaluated only if there is a next change. Better than having nothing, but still won't guard us waiting for a long time with an unprocessed chunk. I will add this logic with a very short `MAX_WAIT_TIME` like 10 seconds, let me know if otherwise.